crates/ra_assists/src/assist_ctx.rs

```diff
@@ -77,20 +77,40 @@ impl<'a, DB: HirDatabase> AssistCtx<'a, DB> {
         f(ctx)
     }
 
-    pub(crate) fn add_assist(
+    pub(crate) fn add_assist_with_single_action(
```

Comment:
Evaluating this direction further, since currently this is the only way of grouping that works for sure.

What if we add an extra `Vec<AssistAction>` field to the `Assist` struct, something like:

```
pub(crate) enum Assist {
    Unresolved { label: AssistLabel },
    Resolved { label: AssistLabel, action: AssistAction, alternative_actions: Vec<AssistAction> },
}
```

and modify the protocol part correspondingly:

```
export async function applySourceChange(ctx: Ctx, change: SourceChange, alternative_changes: SourceChange[] | undefined)
```

?

That way the backward compatibility is preserved (checked by adjusting the protocol part only on a fresh master) and we enforce at least one assist presence on a compile time.